pkg/querybuilder/filter_error_listener.go

```diff
@@ -0,0 +1,235 @@
+) {
+	err := &SyntaxErr{Line: line, Col: column, Msg: msg}
+
+	if tok, ok := off.(antlr.Token); ok {
+		if tok.GetTokenType() == antlr.TokenEOF {
+			err.TokenTxt = "EOF"
+			err.TokenType = tok.GetTokenType()
+		} else {
+			err.TokenTxt = fmt.Sprintf("'%s'", tok.GetText())
+			err.TokenType = tok.GetTokenType()
+		}
+	}
+
+	if p, ok := rec.(antlr.Parser); ok {
+		set := p.GetExpectedTokens()
+
+		// Heuristic: if KEY appears in the expected set *alongside* any literal
+		// value tokens, we assume it stands for a bare value. Otherwise, it stands
+		// for a left‑hand identifier.
+		valueTokens := map[int]struct{}{
+			pGetTokenType(p, "QUOTED_TEXT"): {},
+			pGetTokenType(p, "NUMBER"):      {},
+			pGetTokenType(p, "BOOL"):        {},
+		}
+		hasValueLiterals := false
+		for _, iv := range set.GetIntervals() {
+			for t := iv.Start; t <= iv.Stop; t++ {
+				if _, ok := valueTokens[t]; ok {
+					hasValueLiterals = true
+					break
+				}
+			}
+			if hasValueLiterals {
+				break
+			}
+		}
+
+		uniq := map[string]struct{}{}
+		for _, iv := range set.GetIntervals() {
+			for t := iv.Start; t <= iv.Stop; t++ {
+				sym := tokenName(p, t)
+				if sym == "KEY" {
+					if !hasValueLiterals {
+						uniq["field name (ex: service.name)"] = struct{}{}
+					}
+					continue
+				}
+				if label, ok := prettyToken(p, t); ok {
+					uniq[label] = struct{}{}
+				}
+			}
+		}
+
+		err.Expected = make([]string, 0, len(uniq))
+		for k := range uniq {
+			err.Expected = append(err.Expected, k)
+		}
+		sort.Strings(err.Expected)
+		err.RuleStack = p.GetRuleInvocationStack(nil)
+	}
+
+	l.SyntaxErrors = append(l.SyntaxErrors, err)
+}
+
+func (l *ErrorListener) ReportAmbiguity(
+	rec antlr.Parser,
+	dfa *antlr.DFA,
+	startIdx, stopIdx int,
+	exact bool,
+	ambigAlts *antlr.BitSet,
+	configs *antlr.ATNConfigSet,
+) {
+	if !exact {
+		return
+	}
+	stream := rec.GetTokenStream()
+	txt := textSlice(stream, startIdx, stopIdx)
+	l.Ambigs = append(l.Ambigs, &Ambiguity{
+		Text:   txt,
+		Alts:   ambigAlts.String(),
+		RStack: rec.GetRuleInvocationStack(nil),
+	})
+}
+
+func pGetTokenType(p antlr.Parser, tName string) int {
+	syms := p.GetSymbolicNames()
+	for i, sym := range syms {
+		if sym == tName {
+			return i
+		}
+	}
+	return -1
+}
+
+// tokenName prefers literal > symbolic > numeric.
+func tokenName(p antlr.Parser, tType int) string {
+	lits := p.GetLiteralNames()
+	if tType >= 0 && tType < len(lits) && lits[tType] != "" {
+		return lits[tType]
+	}
+	syms := p.GetSymbolicNames()
+	if tType >= 0 && tType < len(syms) && syms[tType] != "" {
+		return syms[tType]
+	}
+	return fmt.Sprintf("<%d>", tType)
+}
+
+// textSlice pulls raw input text between two token indexes.
+func textSlice(ts antlr.TokenStream, start, stop int) string {
+	var b strings.Builder
+	for i := start; i <= stop && i >= 0; i++ {
+		if tok := ts.Get(i); tok != nil && tok.GetTokenType() != antlr.TokenEOF {
+			b.WriteString(tok.GetText())
+		}
+	}
+	return b.String()
```

Comment:
By adding && i < size to the loop condition, the function would prevent potential index out-of-bounds issues when the stop parameter exceeds the token stream's length.

```
size := ts.Size()  // Get the stream size
    for i := start; i <= stop && i >= 0 && i < size; i++ {
```

Reply:
That should not happen, the stop wouldn't exceed the stream length.